### PR TITLE
Scenes: QueryVariable's query interpolation

### PR DESCRIPTION
**Pull request by dprokop**
Based on https://github.com/grafana/grafana/pull/59553

This PR explores the way to provide data sources with `sceneGraph.interpolate` to correctly resolve QueryVariable's query interpolation. It has a significant footprint of the change to `DataSourceAPI.query` method that now accepts optional `interpolate?: InterpolateFunction` argument that's used as a replacement for `templateSrv.replace` method when provided. This change is also not backwards compatible and would require all data sources to implement the updated `query` interface to allow interpolation via scene graph - see changes in prometheus's implementation in this PR. 
That's a bummer. 

An alternative I want to explore in other PR is to monkey-patch `templateSrv.replace` function when the Scene lifecycle starts and restore the original implementation when the Scene is torn down.

### CODE REVIEW

File: packages/grafana-data/src/types/datasource.ts [hunk 1]
```diff
@@ -245,7 +245,10 @@ abstract class DataSourceApi<
   /**
    * Query for data, and optionally stream results
    */
-  abstract query(request: DataQueryRequest<TQuery>): Promise<DataQueryResponse> | Observable<DataQueryResponse>;
+  abstract query(
+    request: DataQueryRequest<TQuery>,
+    interpolate?: InterpolateFunction
```
Comment (torkelo) [inline packages/grafana-data/src/types/datasource.ts:250]: I think here it should not be optional, implementations should be able to rely on it. But dam that does not work as this function is called a lot FROM plugins, so we have no way to enforce that. 

But having it optional is so messy from the plugin implementation perspective, having to use this but then fallback to something else if it's not there. I think we might need creating brainstorming around this :)

Comment (dprokop) [reply packages/grafana-data/src/types/datasource.ts:250]: Yes - definitelly, just posting this as a possible (quite messy) way to solve the problem.

Let's meet on Monday and talk alternatives through. Tried monkey-patching the templateSrv, but that seems very unreliable.

File: public/app/features/scenes/variables/interpolation/sceneInterpolator.test.ts [hunk 2]
```diff
@@ -51,6 +54,8 @@ describe('sceneInterpolator', () => {
         $variables: new SceneVariableSet({
           variables: [
             new ObjectVariable({
+              // @ts-expect-error
+              type: 'test',
```
Comment (torkelo) [inline public/app/features/scenes/variables/interpolation/sceneInterpolator.test.ts:58]: ?

**Comment by dprokop** [comment]
Closing this, as we figured out an alternative, simpler way to solve the problem 🥳

**Pull request closed without merging**